crates/bevy_pbr/src/material.rs

```diff
@@ -210,11 +209,27 @@ impl<M: SpecializedMaterial> Plugin for MaterialPlugin<M> {
                 .add_render_command::<AlphaMask3d, DrawMaterial<M>>()
                 .init_resource::<MaterialPipeline<M>>()
                 .init_resource::<SpecializedMeshPipelines<MaterialPipeline<M>>>()
+                .add_system_to_stage(RenderStage::Extract, extract_materials::<M>)
                 .add_system_to_stage(RenderStage::Queue, queue_material_meshes::<M>);
         }
     }
 }
 
+fn extract_materials<M: SpecializedMaterial>(
+    mut commands: Commands,
+    query: Query<(Entity, &ComputedVisibility, &Handle<M>)>,
+    mut prev_len: Local<usize>,
+) {
+    let mut materials = Vec::with_capacity(*prev_len);
```

Comment:
This seems overly elaborate: I would prefer to rebase off of https://github.com/bevyengine/bevy/pull/4244 and use the size hint.